cpp/src/stream_compaction/distinct.cu

```diff
@@ -17,28 +17,70 @@
 #include "distinct_helpers.hpp"
 
 #include <cudf/column/column_view.hpp>
+#include <cudf/detail/cuco_helpers.hpp>
 #include <cudf/detail/gather.hpp>
 #include <cudf/detail/iterator.cuh>
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/detail/stream_compaction.hpp>
+#include <cudf/table/experimental/row_operators.cuh>
 #include <cudf/table/table.hpp>
 #include <cudf/table/table_view.hpp>
 #include <cudf/types.hpp>
 
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
 #include <rmm/mr/device/per_device_resource.hpp>
 #include <rmm/resource_ref.hpp>
 
+#include <cuco/static_set.cuh>
 #include <cuda/functional>
 #include <thrust/copy.h>
 #include <thrust/distance.h>
 #include <thrust/iterator/counting_iterator.h>
 #include <thrust/iterator/discard_iterator.h>
 
+#include <atomic>
 #include <utility>
 #include <vector>
 
 namespace cudf {
 namespace detail {
+namespace {
+/**
+ * @brief Invokes the given `func` with desired the row equality and probing method
+ *
+ * @tparam HasNested Flag indicating whether there are nested columns in the input
+ * @tparam Func Type of the helper function doing `distinct` check
+ *
+ * @param compare_nulls Control whether nulls should be compared as equal or not
+ * @param compare_nans Control whether floating-point NaNs values should be compared as equal or not
+ * @param has_nulls Flag indicating whether the input has nulls or not
+ * @param row_equal Self table comparator
+ * @param func The input functor to invoke
+ */
+template <bool HasNested, typename Func>
+rmm::device_uvector<cudf::size_type> dispatch_hash_set(
```

Comment:
I see no reason why this needs to be a separate function. Why can't we have it as a lambda right before calling it?

Reply:
Right, the initial idea is to use this function dispatching CG sizes for performance tuning. Now with the latest experiments, it seems CG size = 1 works the best in all cases thus this dispatching is no longer necessary.

Reply:
We still need a **dispatch** **function** because of the different `row_hasher` types used to contruct the `set` object. The row equality comparators `physical_equality_comparator` and `nan_equal_physical_equality_comparator` will produce different `d_equal` objects which depends on runtime argument `compare_nans`. Hence, this makes it a runtime dispatch than a compile time dispatch, so using lambda's would not simplify.